### PR TITLE
Fix sponge absorbing water across plot borders

### PR DESCRIPTION
**Pull request by traksag**
## Overview

Currently sponges placed on plots also absorb water on the road. E.g. waterlog a plot wall slab, place some water next to it on the road, then place a sponge on your plot next to the waterlogged slab. The water in the waterlogged slab and on road will get removed.

## Description

Use the `SpongeAbsorbEvent` to deal with absorbing water across plot borders (and area borders). This event was added to Bukkit right after it updated to 1.13.

I copied the `StructureGrowEvent` listener and changed it around a bit:
- Removed this seemingly useless part https://github.com/IntellectualSites/PlotSquared/blob/6e7bd0a5365040f9579ff369f11761fd5834c5b3/Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java#L1007-L1024
- Don't cancel water absorption on unowned plots and roads, so staff can use sponges to clear up water there if they so desire.
- Cancel event if no water is going to be absorbed, so the sponge doesn't turn into a wet sponge.

I only tested in a square plot world, I didn't stuff across area borders.

## Checklist

- [x] I included all information required in the sections above
- [x] I tested my changes and approved their functionality
- [x] I ensured my changes do not break other parts of the code
- [x] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/PlotSquared/blob/v6/CONTRIBUTING.md)

### CODE REVIEW

**Review by NotMyFault — APPROVED** [review]
LGTM

File: Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java [hunk 1]
```diff
@@ -1183,4 +1184,37 @@ public void onLeavesDecay(LeavesDecayEvent event) {
...
     }
 
+    @EventHandler(ignoreCancelled = true)
+    public void onSpongeAbsorb(SpongeAbsorbEvent event) {
+        Block sponge = event.getBlock();
+        Location location = BukkitUtil.adapt(sponge.getLocation());
+        PlotArea area = location.getPlotArea();
+        List<org.bukkit.block.BlockState> blocks = event.getBlocks();
+        if (area == null) {
+            for (int i = blocks.size() - 1; i >= 0; i--) {
```
Comment (SirYwell) [inline Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java:1194]: Is there any reason why the list is iterated backwards? If not, a simple `removeIf` would be better probably (and if it needs to be reversed, we could probably use the `ListIterator` instead).

Comment (traksag) [reply Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java:1194]: No particular reason. I copied this loop from https://github.com/IntellectualSites/PlotSquared/blob/6e7bd0a5365040f9579ff369f11761fd5834c5b3/Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java#L982-L987
Similarly for the backwards loop below. Changing it to `removeIf` sounds good to me.

Comment (SirYwell) [reply Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java:1194]: Yeah I guess you can change it then.

**Review by SirYwell — APPROVED** [review]

**Pull request merged**